### DISCUSSION TITLE
[Markdown] Decide what to do about `<div class="index">`

### DISCUSSION THREAD

**Discussion by wbamberg** [mdn/mdn-community#579] [Content] [answered]
This is part of https://github.com/mdn/content/issues/5865.

The CSS documentation sometimes uses markup like `<div class="index">...`. This gets you a multi-column layout, like this:

******
<img width="1034" alt="Screen Shot 2021-07-25 at 3 01 59 PM" src="https://user-images.githubusercontent.com/432915/126914801-d2b1f78d-36d9-4648-9ad1-0f2664e71f68.png">

******

## Where it is used

There are 50 pages under /CSS that use this feature.

<details>
<summary>See the complete list here.</summary>

https://developer.mozilla.org/en-US/docs/Web/CSS/at-rule
https://developer.mozilla.org/en-US/docs/Web/CSS/compositing_and_blending
https://developer.mozilla.org/en-US/docs/Web/CSS/css_animations
https://developer.mozilla.org/en-US/docs/Web/CSS/css_backgrounds_and_borders
https://developer.mozilla.org/en-US/docs/Web/CSS/css_basic_user_interface
https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_alignment/box_alignment_in_flexbox
https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_alignment/box_alignment_in_grid_layout
https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_alignment/box_alignment_in_multi-column_layout
https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_alignment
https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_model
https://developer.mozilla.org/en-US/docs/Web/CSS/css_charsets
https://developer.mozilla.org/en-US/docs/Web/CSS/css_color
https://developer.mozilla.org/en-US/docs/Web/CSS/css_columns
https://developer.mozilla.org/en-US/docs/Web/CSS/css_conditional_rules
https://developer.mozilla.org/en-US/docs/Web/CSS/css_counter_styles
https://developer.mozilla.org/en-US/docs/Web/CSS/css_device_adaptation
https://developer.mozilla.org/en-US/docs/Web/CSS/css_display
https://developer.mozilla.org/en-US/docs/Web/CSS/css_flexible_box_layout
https://developer.mozilla.org/en-US/docs/Web/CSS/css_fonts
https://developer.mozilla.org/en-US/docs/Web/CSS/css_fragmentation
https://developer.mozilla.org/en-US/docs/Web/CSS/css_functions
https://developer.mozilla.org/en-US/docs/Web/CSS/css_generated_content
https://developer.mozilla.org/en-US/docs/Web/CSS/css_grid_layout
https://developer.mozilla.org/en-US/docs/Web/CSS/css_images
https://developer.mozilla.org/en-US/docs/Web/CSS/css_lists_and_counters
https://developer.mozilla.org/en-US/docs/Web/CSS/css_logical_properties
https://developer.mozilla.org/en-US/docs/Web/CSS/css_masking
https://developer.mozilla.org/en-US/docs/Web/CSS/css_miscellaneous
https://developer.mozilla.org/en-US/docs/Web/CSS/css_motion_path
https://developer.mozilla.org/en-US/docs/Web/CSS/css_namespaces
https://developer.mozilla.org/en-US/docs/Web/CSS/css_overflow
https://developer.mozilla.org/en-US/docs/Web/CSS/css_pages
https://developer.mozilla.org/en-US/docs/Web/CSS/css_positioning
https://developer.mozilla.org/en-US/docs/Web/CSS/css_ruby
https://developer.mozilla.org/en-US/docs/Web/CSS/css_scroll_snap_points
https://developer.mozilla.org/en-US/docs/Web/CSS/css_scrollbars
https://developer.mozilla.org/en-US/docs/Web/CSS/css_shapes
https://developer.mozilla.org/en-US/docs/Web/CSS/css_table
https://developer.mozilla.org/en-US/docs/Web/CSS/css_text
https://developer.mozilla.org/en-US/docs/Web/CSS/css_text_decoration
https://developer.mozilla.org/en-US/docs/Web/CSS/css_transforms
https://developer.mozilla.org/en-US/docs/Web/CSS/css_transitions
https://developer.mozilla.org/en-US/docs/Web/CSS/css_variables
https://developer.mozilla.org/en-US/docs/Web/CSS/css_writing_modes
https://developer.mozilla.org/en-US/docs/Web/CSS/cssom_view
https://developer.mozilla.org/en-US/docs/Web/CSS/filter_effects
https://developer.mozilla.org/en-US/docs/Web/CSS/media_queries
https://developer.mozilla.org/en-US/docs/Web/CSS/microsoft_extensions
https://developer.mozilla.org/en-US/docs/Web/CSS/mozilla_extensions
https://developer.mozilla.org/en-US/docs/Web/CSS/webkit_extensions
</details>

Almost all pages that use this are "module pages" - pages that provide an overview of a particular CSS module - where they want to list all the properties in that module in a "Reference" section. For example: https://developer.mozilla.org/en-US/docs/Web/CSS/css_overflow#reference .

There are just a few that aren't:

These three:

https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_alignment/box_alignment_in_flexbox
https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_alignment/box_alignment_in_grid_layout
https://developer.mozilla.org/en-US/docs/Web/CSS/css_box_alignment/box_alignment_in_multi-column_layout

...are subpages of the "CSS Box Alignment" module page, and describe aspects of the module particular to each layout type. I think (but not sure) that the lists of properties they contain should duplicate the list in the main "CSS Box Alignment" module page.

These three:

https://developer.mozilla.org/en-US/docs/Web/CSS/microsoft_extensions
https://developer.mozilla.org/en-US/docs/Web/CSS/mozilla_extensions
https://developer.mozilla.org/en-US/docs/Web/CSS/webkit_extensions

...list proprietary extensions to CSS.

Then there are a few miscellaneous overview pages:

https://developer.mozilla.org/en-US/docs/Web/CSS/at-rule
https://developer.mozilla.org/en-US/docs/Web/CSS/css_functions
https://developer.mozilla.org/en-US/docs/Web/CSS/css_miscellaneous

## What should we do with it?

I can think of three main options:

1. stop supporting it, and list the properties in a simple `<ul>` in one column
2. write a macro that supports it in module pages, and remove it in other places
3. support it by having it included as HTML

### Stop supporting it

I'd vote for this option. Apart from a general principle of wanting to simplify things, there are a couple of other arguments:

* the JS docs (~900 pages) don't need this layout option, and they're generally reckoned to be a particularly high-quality part of MDN. Why is CSS different? Even Web/API only uses it 20 times in 6000 pages.
* in most cases there aren't many items to list, and in those cases multi-column looks quite weird: https://developer.mozilla.org/en-US/docs/Web/CSS/css_lists_and_counters#properties. A single column would be better here.
* even where there are many items https://developer.mozilla.org/en-US/docs/Web/CSS/css_backgrounds_and_borders#css_properties it's not clear to me that a multi-column list is easier to scan, because you have to think about which column the item you are looking for is in.

### Use a macro

I believe that this would be reasonably practical. mdn/data has a "groups" item for each property (e.g. https://github.com/mdn/data/blob/master/css/properties.json#L2046), so we could write a macro that takes a group name as a parameter and renders a list of the properties in that group. This would be better than the current situation where the lists are maintained by hand.

Because the macro emits HTML, it can emit `<div class="index">` and get the fancy layout.

We've started to talk about [using webref instead of mdn/data](https://github.com/openwebdocs/project/issues/44), but we could probably switch to using webref easily enough in the future.

### Support it as HTML

This would be my least favourite option.

**Comment by hamishwillee** [comment]
My 2 bits:
- I prefer single column list for small lists, and I'd change them irrespective of what we decide to do about this syntax. 
- For those larger lists I disagree that they are less useful or odd looking. It is frustrating to have very long lists, particularly if there is content on the other side you are interested in.
- I tend towards the idea of using macros to generate this kind of thing - either from data or subdirectory levels. 
- Agree HTML not a good plan. 

I suspect JS must have large lists of properties too. Are there any cases to compare against?

**Reply by teoli2003** [reply]
I agree with Hamish here:
- HTML is not a good plan
- Small lists should be single column (E.g. https://developer.mozilla.org/en-US/docs/Glossary/Flex_item or https://developer.mozilla.org/en-US/docs/Web/API/Sensor and most other API overview pages)
- A _few_ long lists are better in multicolumn.

As the name of the class says, the `index` class was created for large indices and not for creating multicolumn lists.
`{{CSS_Ref}}`, `{{ListGroups}}`, and `{{APIListAlpha}}` are a few macros using them. Most of the time these macros are used only once. They exists as indices need to be automatically generated to be kept up-to-date.

If we remove the use of `class="index"` for small lists, and keep it in these macros, I think we will be left with very few  cases. We could then discuss what to do with each of them.

Here is one example: https://developer.mozilla.org/en-US/docs/Web/API/PerformanceEventTiming

In other words, lets keep `index` for actual indices (and generated in macros); use simple lists or definition lists for the _small list_ use case; and handle case by case the remaining ones.

**Reply by wbamberg** [reply]
Thanks for the views!

Cases inside macros are not a problem for Markdown of course. The cases listed here are all outside macros, and the ones with the longest lists are the proprietary CSS ones:

https://developer.mozilla.org/en-US/docs/Web/CSS/microsoft_extensions
https://developer.mozilla.org/en-US/docs/Web/CSS/mozilla_extensions
https://developer.mozilla.org/en-US/docs/Web/CSS/webkit_extensions

After that are some of  the bigger modules, like:

https://developer.mozilla.org/en-US/docs/Web/CSS/css_backgrounds_and_borders

There's also:

https://developer.mozilla.org/en-US/docs/Web/CSS/css_functions

...but that's weird because it lists all the functions twice, once separate by area and once in a single big index. I question the value of that.

So for pages like these, the question is still, do we live with single-col or write a macro? One complicating factor is that while it would probably be quite easy to write a macro that lists all properties belonging to a given (mdn/data) group, that won't work for much of https://developer.mozilla.org/en-US/docs/Web/CSS/mozilla_extensions , which wants to list all kinds of things in this way (including, for example, lists of values that various properties can accept).

**Reply by Ryuno-Ki** [reply]
I vote for single-column.
We can avoid issues like https://github.com/mdn/mdn-minimalist/issues/682 this way easily.

**Reply by wbamberg** [reply]
Yeah, there are definitely issues with the way multicol is implemented on MDN at the moment (also here: https://developer.mozilla.org/en-US/docs/Web/CSS/Mozilla_Extensions#mozilla-only_properties_and_pseudo-classes_avoid_using_on_websites the layout is broken and the whole page scrolls). Of course these kinds of things are fixable but I'm always on team keep it simple.

**Comment by rachelandrew** [comment]
I'd be happy with these being a single column. I also wanted to remove the duplication on that CSS Functions page a while ago, I can't remember why that suggestion was outvoted, but I have never thought the index that useful and it's a pain to keep updated.

Background and Borders could probably do with an update, it doesn't include the flow relative mappings, which it probably should. I wonder whether I could reformat that list into groups, thus making it not just a long list anyway.

If the above sounds ok, I could deal with those by removing the duplication on CSS Functions, and reformatting backgrounds and borders to be hopefully also more useful.

I'm not sure whatr to suggest with those extensions pages, there was another issue raised about those yesterday too. Maybe they need looking at more generally.

**Reply by wbamberg** [reply]
Thanks Rachel! 

> If the above sounds ok

That all sounds great to me!

> I'm not sure whatr to suggest with those extensions pages, there was another issue raised about those yesterday too. Maybe they need looking at more generally.

Yes, I'm not sure either. I'm not sure they are terribly useful pages, and as you said in the other bug they don't seem fully up to date. Maybe it would be OK to single-column them for the time being?

**Comment by hamishwillee** [comment]
No need to guess. [From this PR](https://github.com/mdn/content/pull/7361#issuecomment-887875275) see what single-column would look like for some of these cases: 
- <https://pr7361.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/Microsoft_Extensions>
- <https://pr7361.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/Mozilla_Extensions>
- <https://pr7361.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/CSS_Lists_and_Counters>
- <https://pr7361.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/CSS_Backgrounds_and_Borders>

My take on this is that the compact view is better, but I can live with the single column (obviously some tidy up required - I just removed the divs)

**Reply by wbamberg** [reply]
Thank you Hamish! I guess it's subjective and I am biased by wanting to keep things simple, but I think it looks all right. And there are places where single-column is much more scannable than multi, such as:

https://pr7361.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/Mozilla_Extensions#formerly_proprietary_properties_that_are_now_standard

versus 

https://developer.mozilla.org/en-US/docs/Web/CSS/Mozilla_Extensions#formerly_proprietary_properties_that_are_now_standard

**Reply by hamishwillee** [reply]
I think you can go ahead with your plan - option 1 to remove the class that forces lists to be multi-column. Certainly the few cases where I prefer the compact view are not worth the complexity, and might be addressed on case by case basis.

**Comment by wbamberg** [comment, marked as answer]
Thanks for everyone's thought on this! I just filed https://github.com/mdn/content/pull/7726 to remove `<div class="index">` from the CSS documentation, and Hamish merged. So we can also close this discussion.

I do think for the module overview pages a macro is the correct option, but as far as Markdown is concerned removing them is just as good.

**Reply by hamishwillee** [reply]
It certainly looks either better or as good for all the cases you fixed in 7726. I checked :-)

**Comment by pepelsbey** [comment]
Closing as answered/settled.
